protocol/shannon/fullnode_session_rollover.go

```diff
@@ -0,0 +1,167 @@
+// - sessionRolloverEnd: End of the rollover window (session end height + grace period)
+// These boundaries are only updated when the current rollover period has ended,
+// preventing frequent recalculation and ensuring stable rollover detection.
+//
+// All access to this state is protected by rolloverStateMu for thread safety,
+// allowing safe concurrent access from multiple goroutines.
+type sessionRolloverState struct {
+	currentBlockHeight   int64 // Latest block height from the blockchain
+	sessionRolloverStart int64 // Start height of the rollover window
+	sessionRolloverEnd   int64 // End height of the rollover window
+	isInSessionRollover  bool  // Cached rollover status (true = in rollover period)
+
+	rolloverStateMu sync.RWMutex // Protects all fields above
+}
+
+// getSessionRolloverState returns whether we're currently in a session rollover period.
+// Thread-safe.
+func (lfn *LazyFullNode) getSessionRolloverState() bool {
+	lfn.rolloverState.rolloverStateMu.RLock()
+	defer lfn.rolloverState.rolloverStateMu.RUnlock()
+	return lfn.rolloverState.isInSessionRollover
+}
+
+// startSessionRolloverMonitoring starts background monitoring for session rollovers.
+// Called automatically when LazyFullNode is created.
+func (lfn *LazyFullNode) startSessionRolloverMonitoring() {
+	lfn.logger.Info().
+		Dur("check_interval", blockCheckInterval).
+		Int("grace_period_blocks", sessionRolloverGracePeriodBlocks).
+		Msg("Starting session rollover monitoring")
+
+	go lfn.monitorLoop()
+}
+
+// monitorLoop continuously checks block height to detect session rollovers
+func (lfn *LazyFullNode) monitorLoop() {
+	ticker := time.NewTicker(blockCheckInterval)
+	defer ticker.Stop()
+
+	for range ticker.C {
+		lfn.updateBlockHeight()
+	}
+}
+
+// updateSessionValues updates rollover boundaries when we fetch a new session.
+// Called from GetSession() to keep rollover monitoring current.
+// Only updates rollover boundaries if the current rollover period has ended.
+func (lfn *LazyFullNode) updateSessionValues(session sessiontypes.Session) {
+	if session.Header == nil {
+		lfn.logger.Warn().Msg("Session header is nil, cannot update session values")
+		return
+	}
+
+	lfn.rolloverState.rolloverStateMu.Lock()
+	defer lfn.rolloverState.rolloverStateMu.Unlock()
+
+	sessionEndHeight := session.Header.SessionEndBlockHeight
+	currentBlockHeight := lfn.rolloverState.currentBlockHeight
+
+	// Only update rollover boundaries if current rollover period has ended
+	// or if we don't have rollover boundaries set yet
+	if lfn.rolloverState.sessionRolloverEnd == 0 || currentBlockHeight > lfn.rolloverState.sessionRolloverEnd {
+		newRolloverStart := sessionEndHeight - 1
+		newRolloverEnd := sessionEndHeight + sessionRolloverGracePeriodBlocks
+
+		oldRolloverStart := lfn.rolloverState.sessionRolloverStart
+
+		// Update rollover boundaries and recalculate rollover status
+		lfn.rolloverState.sessionRolloverStart = newRolloverStart
```

Comment:
Moving all these changes into a method of the rolloverState struct would clarify things quite a bit.